controller/src/main/java/io/pravega/controller/server/eventProcessor/CommitEventProcessor.java

```diff
@@ -77,45 +81,83 @@ public CommitEventProcessor(final StreamMetadataStore streamMetadataStore,
+            }
+        }, executor).join();
     }
 
-    private CompletableFuture<Controller.TxnStatus> notifyCommitToHost(final String scope, final String stream, final int segmentNumber, final UUID txId) {
-        final long retryInitialDelay = 100;
-        final int retryMultiplier = 10;
-        final int retryMaxAttempts = 100;
-        final long retryMaxDelay = 100000;
+    private CompletableFuture<Void> completeCommit(final String scope,
+                                                   final String stream,
+                                                   final int epoch,
+                                                   final List<Integer> segments,
+                                                   final UUID txnId,
+                                                   final OperationContext context) {
+        return notifyCommitToHost(scope, stream, segments, txnId).thenComposeAsync(x ->
+                streamMetadataStore.commitTransaction(scope, stream, epoch, txnId, context, executor), executor)
+                .thenApply(x -> null);
+    }
+
+    private CompletableFuture<Void> postponeCommitEvent(CommitEvent event) {
+        return Retry.indefinitelyWithExpBackoff("Error writing event back into CommitStream")
+                .runAsync(() -> writeEvent(event), executor);
+    }
+
+    private CompletableFuture<Void> writeEvent(CommitEvent event) {
+        UUID txnId = event.getTxid();
+        log.debug("Transaction {}, pushing back CommitEvent to commitStream", txnId);
+        AckFuture future = this.getSelfWriter().write(event);
+        CompletableFuture<AckFuture> writeComplete = new CompletableFuture<>();
+        future.addListener(() -> writeComplete.complete(future), executor);
+        return writeComplete.thenApplyAsync(ackFuture -> {
+            try {
+                // ackFuture is complete by now, so we can do a get without blocking
+                ackFuture.get();
+                log.debug("Transaction {}, sent request to commitStream", txnId);
+                return null;
+            } catch (InterruptedException e) {
+                log.warn("Transaction {}, unexpected interruption sending event to commitStream. Retrying...", txnId);
+                throw new WriteFailedException(e);
+            } catch (ExecutionException e) {
+                Throwable realException = ExceptionHelpers.getRealException(e);
+                log.warn("Transaction {}, failed sending event to commitStream. Retrying...", txnId);
+                throw new WriteFailedException(realException);
+            }
+        }, executor);
+    }
+
+    private CompletableFuture<Void> notifyCommitToHost(final String scope, final String stream,
+                                                       final List<Integer> segments, final UUID txnId) {
+        return FutureHelpers.allOf(segments.stream()
+                .parallel()
+                .map(segment -> notifyCommitToHost(scope, stream, segment, txnId))
```

Comment:
Could you remind me what happens in the case the controller instance crashes in the middle of notifying the commit to the segment store servers?